### PR TITLE
bump liquibase to 4.12.0

### PR DESCRIPTION
**Pull request by mshima**
Fix https://github.com/jhipster/generator-jhipster/issues/18071
---

Please make sure the below checklist is followed for Pull Requests.

- [ ] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [ ] Tests are added where necessary
- [ ] The JDL part is updated if necessary
- [ ] [jhipster-online](https://github.com/jhipster/jhipster-online) is updated if necessary
- [ ] Documentation is added/updated where necessary
- [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (below reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

**Comment by mshima** [comment]
not released yet.

**Comment by DanielFran** [comment]
@mshima we need to wait for liquibase-hibernate update too...

File: generators/server/templates/build.gradle.ejs [hunk 1]
```diff
@@ -250,12 +250,6 @@ configurations {
             // Inherited version from Spring Boot can't be used because of regressions:
             // To be removed as soon as spring-boot use the same version
             force 'org.liquibase:liquibase-core:<%= LIQUIBASE_VERSION %>'
```
Comment (mshima) [inline generators/server/templates/build.gradle.ejs:252]:
```suggestion
            force 'org.liquibase:liquibase-core:<%= LIQUIBASE_VERSION %>'
<%_ if (devDatabaseTypeH2Any && reactive) { _%>
            force 'com.h2database:h2:<%= H2_VERSION %>'
            force 'io.r2dbc:r2dbc-h2:<%= H2_R2DBC_VERSION %>'
<%_ } _%>
```

File: generators/server/templates/pom.xml.ejs [hunk 2]
```diff
@@ -102,12 +102,6 @@
         <liquibase.version><%= LIQUIBASE_VERSION %></liquibase.version>
   <%_ if (!reactive) { _%>
         <liquibase-hibernate5.version><%= LIQUIBASE_VERSION %></liquibase-hibernate5.version>
```
Comment (mshima) [inline generators/server/templates/pom.xml.ejs:104]:
```suggestion
        <liquibase-hibernate5.version><%= LIQUIBASE_VERSION %></liquibase-hibernate5.version>
  <%_ } _%>
  <%_ if (devDatabaseTypeH2Any && reactive) { _%>
        <h2.version><%= H2_VERSION %></h2.version>
        <h2-r2dbc.version><%= H2_R2DBC_VERSION %></h2-r2dbc.version>
```

File: generators/server/templates/pom.xml.ejs [hunk 3]
```diff
@@ -1581,13 +1579,11 @@
                 <dependency>
                     <groupId>com.h2database</groupId>
                     <artifactId>h2</artifactId>
```
Comment (mshima) [inline generators/server/templates/pom.xml.ejs:1581]:
```suggestion
                    <artifactId>h2</artifactId>
<%_ if (reactive) { _%>
                    <version>${h2.version}</version>
<%_ } _%>
```

File: generators/server/templates/pom.xml.ejs [hunk 4]
```diff
@@ -1581,13 +1579,11 @@
...
-                    <version>${h2.version}</version>
                 </dependency>
   <%_ if (reactive) { _%>
                 <dependency>
                     <groupId>io.r2dbc</groupId>
                     <artifactId>r2dbc-h2</artifactId>
```
Comment (mshima) [inline generators/server/templates/pom.xml.ejs:1586]:
```suggestion
                    <artifactId>r2dbc-h2</artifactId>
                    <version>${h2-r2dbc.version}</version>
```

File: generators/server/templates/pom.xml.ejs [hunk 5]
```diff
@@ -103,11 +103,9 @@
   <%_ if (!reactive) { _%>
         <liquibase-hibernate5.version><%= LIQUIBASE_VERSION %></liquibase-hibernate5.version>
   <%_ } _%>
-  <%_ if (devDatabaseTypeH2Any) { _%>
```
Comment (DanielFran) [inline generators/server/templates/pom.xml.ejs:106]: There is a reference to h2.version here: https://github.com/jhipster/generator-jhipster/blob/8cb356615a28c21e128f1c6f3fab8c8faf65f8b7/generators/server/templates/pom.xml.ejs#L1779

File: generators/server/templates/build.gradle.ejs [hunk 6]
```diff
@@ -250,11 +250,9 @@ configurations {
...
-<%_ if (devDatabaseTypeH2Any) { _%>
+<%_ if (devDatabaseTypeH2Any && reactive) { _%>
```
Comment (Tcharl) [inline generators/server/templates/build.gradle.ejs:253]: No need for a driver for imperative?

File: generators/server/templates/src/main/resources/config/application-dev.yml.ejs [hunk 7]
```diff
@@ -138,11 +138,11 @@ spring:
     username: SA
     password: yourStrong(!)Password
   <%_ } else if (devDatabaseTypeH2Disk) { _%>
-    url: <%- getJDBCUrl(devDatabaseType, { databaseName: lowercaseBaseName, localDirectory: `./${BUILD_DIR}h2db/db` }) %>
+    url: <%- getJDBCUrl(devDatabaseType, { prodDatabaseType, databaseName: lowercaseBaseName, localDirectory: `./${BUILD_DIR}h2db/db` }) %>
```
Comment (Tcharl) [inline generators/server/templates/src/main/resources/config/application-dev.yml.ejs:141]: Please add a specific option (`dbCompatMode: <% prodDatabaseType %>`).

**Comment by DanielFran** [comment]
liquibase-core and liquibase-hibernate will be synchronized only for next version 4.13.0

**Comment by DanielFran** [comment]
@mshima liquibase-hibernate v4.12.0 has been released

File: generators/server/templates/build.gradle.ejs (hunk 6 above)
Comment (mshima) [reply generators/server/templates/build.gradle.ejs:253]: It’s overriding the version for reactive only

File: generators/server/templates/src/main/resources/config/application-dev.yml.ejs [hunk 8]
```diff
@@ -207,9 +207,9 @@ spring:
   <%_ } else if (devDatabaseTypeOracle) { _%>
     url: <%- getJDBCUrl(devDatabaseType, { databaseName: 'xe', hostname: 'localhost' }) %>
   <%_ } else if (devDatabaseTypeH2Disk) { _%>
-    url: <%- getJDBCUrl(devDatabaseType, { databaseName: lowercaseBaseName, localDirectory: `./${BUILD_DIR}h2db/db` }) %>
+    url: <%- getJDBCUrl(devDatabaseType, { prodDatabaseType, databaseName: lowercaseBaseName, localDirectory: `./${BUILD_DIR}h2db/db` }) %>
   <%_ } else if (devDatabaseTypeH2Memory) { _%>
-    url: <%- getJDBCUrl(devDatabaseType, { databaseName: lowercaseBaseName }) %>
+    url: <%- getJDBCUrl(devDatabaseType, { prodDatabaseType, databaseName: lowercaseBaseName }) %>
```
Comment (mshima) [inline generators/server/templates/src/main/resources/config/application-dev.yml.ejs:212]:
```suggestion
    url: <%- getJDBCUrl(devDatabaseType, { databaseName: lowercaseBaseName }) %>
```
```suggestion
    url: <%- getJDBCUrl(devDatabaseType, { prodDatabaseType, databaseName: lowercaseBaseName }) %>
```

File: generators/generator-constants.js [hunk 9]
```diff
@@ -36,16 +36,16 @@ const OPENAPI_GENERATOR_CLI_VERSION = '2.5.1';
...
-const LIQUIBASE_VERSION = '4.11.0';
+const LIQUIBASE_VERSION = '4.12.0';
 const LIQUIBASE_DTD_VERSION = LIQUIBASE_VERSION.split('.', 3).slice(0, 2).join('.');
 const HIBERNATE_VERSION = '5.6.9.Final';
 const JACOCO_VERSION = '0.8.8';
 const JACKSON_DATABIND_NULLABLE_VERSION = '0.2.2';
 const JIB_VERSION = '3.2.1';
 // TODO upgrade to h2 v2 once liquibase issues has been addressed
-const H2_VERSION = '1.4.200';
+const H2_VERSION = '2.1.214';
```
Comment (mshima) [inline generators/generator-constants.js:46]:
```suggestion
const H2_VERSION = '1.4.200';
```

File: generators/generator-constants.js [hunk 10]
```diff
@@ -36,16 +36,16 @@ const OPENAPI_GENERATOR_CLI_VERSION = '2.5.1';
...
 // Version 0.9.x is not compatible with h2 v1
-const H2_R2DBC_VERSION = '0.8.5.RELEASE';
+const H2_R2DBC_VERSION = '0.9.1.RELEASE';
```
Comment (mshima) [inline generators/generator-constants.js:48]:
```suggestion
const H2_R2DBC_VERSION = '0.8.5.RELEASE';
```

File: generators/server/templates/sql/common/src/test/resources/config/application-testdev.yml.ejs [hunk 11]
```diff
@@ -49,9 +49,9 @@ spring:
     password:
   liquibase:
     <%_ if (devDatabaseTypeH2Memory) { _%>
-    url: <%- getJDBCUrl(devDatabaseType, { databaseName: lowercaseBaseName, itests: true }) %>
+    url: <%- getJDBCUrl(devDatabaseType, { prodDatabaseType, databaseName: lowercaseBaseName, itests: true }) %>
```
Comment (mshima) [inline generators/server/templates/sql/common/src/test/resources/config/application-testdev.yml.ejs:52]:
```suggestion
    url: <%- getJDBCUrl(devDatabaseType, { databaseName: lowercaseBaseName, itests: true }) %>
```

File: generators/server/templates/sql/common/src/test/resources/config/application-testdev.yml.ejs [hunk 12]
```diff
@@ -49,9 +49,9 @@ spring:
...
     <%_ } else { _%>
-    url: <%- getJDBCUrl(devDatabaseType, { databaseName: lowercaseBaseName, localDirectory: `./${BUILD_DIR}h2db/testdb` }) %>
+    url: <%- getJDBCUrl(devDatabaseType, { prodDatabaseType, databaseName: lowercaseBaseName, localDirectory: `./${BUILD_DIR}h2db/testdb` }) %>
```
Comment (mshima) [inline generators/server/templates/sql/common/src/test/resources/config/application-testdev.yml.ejs:54]:
```suggestion
    url: <%- getJDBCUrl(devDatabaseType, { databaseName: lowercaseBaseName, localDirectory: `./${BUILD_DIR}h2db/testdb` }) %>
```

File: generators/server/templates/src/main/resources/config/application-dev.yml.ejs (hunk 7 above)
Comment (mshima) [inline generators/server/templates/src/main/resources/config/application-dev.yml.ejs:141]:
```suggestion
    url: <%- getJDBCUrl(devDatabaseType, { databaseName: lowercaseBaseName, localDirectory: `./${BUILD_DIR}h2db/db` }) %>
```

File: generators/server/templates/src/main/resources/config/application-dev.yml.ejs [hunk 13]
```diff
@@ -138,11 +138,11 @@ spring:
     username: SA
     password: yourStrong(!)Password
   <%_ } else if (devDatabaseTypeH2Disk) { _%>
-    url: <%- getJDBCUrl(devDatabaseType, { databaseName: lowercaseBaseName, localDirectory: `./${BUILD_DIR}h2db/db` }) %>
+    url: <%- getJDBCUrl(devDatabaseType, { prodDatabaseType, databaseName: lowercaseBaseName, localDirectory: `./${BUILD_DIR}h2db/db` }) %>
     username: <%= baseName %>
     password:
   <%_ } else if (devDatabaseTypeH2Memory) { _%>
-    url: <%- getJDBCUrl(devDatabaseType, { databaseName: lowercaseBaseName }) %>
+    url: <%- getJDBCUrl(devDatabaseType, { prodDatabaseType, databaseName: lowercaseBaseName }) %>
```
Comment (mshima) [inline generators/server/templates/src/main/resources/config/application-dev.yml.ejs:145]:
```suggestion
    url: <%- getJDBCUrl(devDatabaseType, { databaseName: lowercaseBaseName }) %>
```

File: generators/server/templates/src/main/resources/config/application-dev.yml.ejs [hunk 14]
```diff
@@ -207,9 +207,9 @@ spring:
   <%_ } else if (devDatabaseTypeOracle) { _%>
     url: <%- getJDBCUrl(devDatabaseType, { databaseName: 'xe', hostname: 'localhost' }) %>
   <%_ } else if (devDatabaseTypeH2Disk) { _%>
-    url: <%- getJDBCUrl(devDatabaseType, { databaseName: lowercaseBaseName, localDirectory: `./${BUILD_DIR}h2db/db` }) %>
+    url: <%- getJDBCUrl(devDatabaseType, { prodDatabaseType, databaseName: lowercaseBaseName, localDirectory: `./${BUILD_DIR}h2db/db` }) %>
```
Comment (mshima) [inline generators/server/templates/src/main/resources/config/application-dev.yml.ejs:210]:
```suggestion
    url: <%- getJDBCUrl(devDatabaseType, { databaseName: lowercaseBaseName, localDirectory: `./${BUILD_DIR}h2db/db` }) %>
```

**Comment by mshima** [comment]
My guess the h2 bug is https://github.com/liquibase/liquibase/issues/2306

File: generators/server/templates/build.gradle.ejs (hunk 6 above)
Comment (mshima) [inline generators/server/templates/build.gradle.ejs:253]:
```suggestion
<%_ if (devDatabaseTypeH2Any && prodDatabaseTypeMysql) { _%>
```

File: generators/server/templates/pom.xml.ejs [hunk 15]
```diff
@@ -103,11 +103,9 @@
...
+  <%_ if (devDatabaseTypeH2Any && reactive) { _%>
```
Comment (mshima) [inline generators/server/templates/pom.xml.ejs:106]:
```suggestion
  <%_ if (devDatabaseTypeH2Any && prodDatabaseTypeMysql) { _%>
```

File: generators/server/templates/pom.xml.ejs [hunk 16]
```diff
@@ -1581,7 +1579,9 @@
                 <dependency>
                     <groupId>com.h2database</groupId>
                     <artifactId>h2</artifactId>
+<%_ if (reactive) { _%>
```
Comment (mshima) [inline generators/server/templates/pom.xml.ejs:1582]:
```suggestion
<%_ if (prodDatabaseTypeMysql) { _%>
```

File: generators/server/templates/pom.xml.ejs [hunk 17]
```diff
@@ -1776,7 +1776,9 @@
                                 <dependency>
                                     <groupId>com.h2database</groupId>
                                     <artifactId>h2</artifactId>
+    <%_ if (reactive) { _%>
```
Comment (mshima) [inline generators/server/templates/pom.xml.ejs:1779]:
```suggestion
    <%_ if (prodDatabaseTypeMysql) { _%>
```

**Review by mshima — COMMENTED** [review]
Keep h2 v1 when using mysql as prod database.

File: generators/server/templates/build.gradle.ejs [hunk 18]
```diff
@@ -250,7 +250,7 @@ configurations {
             // Inherited version from Spring Boot can't be used because of regressions:
             // To be removed as soon as spring-boot use the same version
             force 'org.liquibase:liquibase-core:<%= LIQUIBASE_VERSION %>'
-<%_ if (devDatabaseTypeH2Any) { _%>
+<%_ if (devDatabaseTypeH2Any && prodDatabaseTypeMysql) { _%>
```
Comment (mshima) [inline generators/server/templates/build.gradle.ejs:253]:
```suggestion
<%_ if (devDatabaseTypeH2Any && (prodDatabaseTypeMysql || reactive)) { _%>
```

File: generators/server/templates/pom.xml.ejs [hunk 19]
```diff
@@ -103,7 +103,7 @@
   <%_ if (!reactive) { _%>
         <liquibase-hibernate5.version><%= LIQUIBASE_VERSION %></liquibase-hibernate5.version>
   <%_ } _%>
-  <%_ if (devDatabaseTypeH2Any) { _%>
+  <%_ if (devDatabaseTypeH2Any && prodDatabaseTypeMysql) { _%>
```
Comment (mshima) [inline generators/server/templates/pom.xml.ejs:106]:
```suggestion
  <%_ if (devDatabaseTypeH2Any && (prodDatabaseTypeMysql || reactive)) { _%>
```

File: generators/server/templates/pom.xml.ejs [hunk 20]
```diff
@@ -1581,13 +1581,17 @@
                 <dependency>
                     <groupId>com.h2database</groupId>
                     <artifactId>h2</artifactId>
+  <%_ if (prodDatabaseTypeMysql) { _%>
```
Comment (mshima) [inline generators/server/templates/pom.xml.ejs:1584]:
```suggestion
  <%_ if (prodDatabaseTypeMysql || reactive) { _%>
```

**Comment by mshima** [comment]
Use h2 v1 whenever `liquibaseAutoIncrement` is used (reactive || mysql):

https://github.com/jhipster/generator-jhipster/blob/d048756de18718433196c78c7703c4f37cfdeafd/utils/field.js#L282-L298

**Comment by mshima** [comment]
Please squash the PR

**Pull request merged**